### PR TITLE
Feature/35/simple nlp algorithm

### PR DESCRIPTION
**Pull request by grillmann**

### CODE REVIEW

File: server/controllers/nlp.js [hunk 1]
```diff
@@ -47,12 +208,76 @@ exports.handleInput = function (req, res) {
...
 
+        let lowerCaseInput = req.body.input.toLowerCase();
+
+        var tokenizer = new natural.WordTokenizer();
+        let tokenizedInput = tokenizer.tokenize(lowerCaseInput);
+
+
+        let initState = {
+            tokens: tokenizedInput,
+            layer: 0,
```
Comment (syxolk) [inline server/controllers/nlp.js]: What is a layer?

Comment (grillmann) [reply server/controllers/nlp.js]: Layers are the different types of Expressions we search for and then label the words. So for example charttype Column, Connector or operations

File: server/controllers/nlp.js [hunk 2]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+        // this layer matches for this token
+        state.tokens[state.currentToken] = {type: "Operation", value: currentToken}
+        // this layer matches for this token
+        state.currentToken++;
+        state.layer = 0;
+        findCommand(state);
+    } else {
+        // this layer does not match
+        nextActionAfterNotMatched(state);
     }
```
Comment (sebmarste) [inline server/controllers/nlp.js]: can we extract the whole if to a nextAction(matched) call

File: server/controllers/nlp.js [hunk 3]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+        });
+    },
+    plotPieChartOfColumn: function (dataset, parameters, callback) {
+        let column = parameters[0];
+        // SELECT <column> FROM <dataset>
+        bookshelf.Model.extend({tableName: dataset}).fetchAll({columns: [column]}).then(data => {
+            var entries = data.map((value) => value.attributes[column]);
+            var values = [];
+            var labels = [];
+            _.forEach(entries, function (entry) {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: maybe use entries.forEach(entry => ...) here

Comment (sebmarste) [reply server/controllers/nlp.js]: generally i would prefer array.function(...) to _.function(array, ...), what do you think about that?

File: server/controllers/nlp.js [hunk 4]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+}
+
+function extractConnector(state) {
+
+}
+
+let searchFunction = [
+    extractColumn,
+    extractChartType,
+    extractOperation
```
Comment (sebmarste) [inline server/controllers/nlp.js]: maybe change the order here: 
1) operation
2) chartType
3) column

File: server/controllers/nlp.js [hunk 5]
```diff
@@ -47,12 +208,76 @@ exports.handleInput = function (req, res) {
...
+
+        var tokenizer = new natural.WordTokenizer();
+        let tokenizedInput = tokenizer.tokenize(lowerCaseInput);
+
+
+        let initState = {
+            tokens: tokenizedInput,
+            layer: 0,
+            currentToken: 0,
+            callback: state => {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: extract this function

File: server/controllers/nlp.js [hunk 6]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+                x: data.map((value) => value.attributes[column]),
+                type: 'scatter'
+            }]
+        ))
+    },
+    plotScatterOfTwoColumns: function (dataset, parameters, callback) {
+        let column1 = parameters[0];
+        let column2 = parameters[1];
+
+        // SELECT <column> FROM <dataset>
```
Comment (sebmarste) [inline server/controllers/nlp.js]: copy past error, or?

File: server/controllers/nlp.js [hunk 7]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+                    y: data.map((value) => value.attributes[column2]),
+                    mode: 'markers',
+                    type: 'scatter'
+                }]
+            )
+        });
+    },
+    plotPieChartOfColumn: function (dataset, parameters, callback) {
+        let column = parameters[0];
+        // SELECT <column> FROM <dataset>
```
Comment (sebmarste) [inline server/controllers/nlp.js]: this should be updated as well

File: server/controllers/nlp.js (hunk 2 above)
Comment (sebmarste) [reply server/controllers/nlp.js]: or maybe even merge it with findCommand to a gereric controller function

File: server/controllers/nlp.js [hunk 8]
```diff
@@ -47,12 +208,76 @@ exports.handleInput = function (req, res) {
...
+                    numberMatches: 0,
+                    function: "",
+                    functionParameter: []
+                };
+                _.forEach(commands, function (command) {
+                    numberMatches = 0;
+                    numberColumns = 0;
+                    columnsArray = [];
+                    _.forEach(state.tokens, function (token) {
+                        if (token.type === "Column") {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: could we use a static section where we define such strings only once and use them as reference from then on?
```
let static = {
   column: "Column",
   ...
}

...

if(token.type === static.column) {
...
```

File: server/data/commands.json [hunk 9]
```diff
@@ -25,4 +112,7 @@
       }
     ]
   }
+
+
+
```
Comment (sebmarste) [inline server/data/commands.json]: we should think about how to restructure this document correclty

File: server/controllers/nlp.js [hunk 10]
```diff
@@ -47,12 +208,76 @@ exports.handleInput = function (req, res) {
...
+                        }
+                    });
+                    if(numberColumns === command.parameters.numberColumns){
+                        numberMatches++;
+                    }
+                    if(numberMatches > bestMatched.numberMatches){
+                        bestMatched.numberMatches = numberMatches;
+                        bestMatched.function = command.function;
+                        _.forEach(command.functionParameters, function(param, index){
+                            if(param === "Column"){
```
Comment (r-sukanya) [inline server/controllers/nlp.js]: what is the purpose of this ? it will always be true as command.functionParameters = "column"

Comment (grillmann) [reply server/controllers/nlp.js]: Well, it depends what we need to tansform the data. For now it just uses columns, but maybe later it needs also other input like if we should plot percent or total count.

File: server/controllers/nlp.js (hunk 6 above)
Comment (grillmann) [reply server/controllers/nlp.js]: Do you mean that it should be <column1, column2> in the comment?

File: server/controllers/nlp.js (hunk 7 above)
Comment (grillmann) [reply server/controllers/nlp.js]: what should be updated?

File: server/controllers/nlp.js (hunk 8 above)
Comment (grillmann) [reply server/controllers/nlp.js]: good idea, thanks!

File: server/controllers/nlp.js (hunk 6 above)
Comment (sebmarste) [reply server/controllers/nlp.js]: `// SELECT <column1>, <column2> FROM <dataset>`

File: server/controllers/nlp.js (hunk 7 above)
Comment (sebmarste) [reply server/controllers/nlp.js]: you are doing a groupting and counting here, right? can we do this directly on the db? if yes, please update the sql "query" as well, if not, how can we document the data transformations that happen here?

Comment (sebmarste) [reply server/controllers/nlp.js]: a query would look like this `SELECT count(*), <column> FROM <dataset> GROUP BY <column>`

File: server/controllers/nlp.js [hunk 11]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
-        function: 'plotHistogramOfColumn',
-        parameters: [
-            'age'
-        ]
+
+function extractOperation(state) {
+    let possibleOperations = ["plot", "make", "draw", "select"];
+    var currentToken = state.tokens[state.currentToken];
+
+    var result = _.includes(possibleOperations, currentToken);
```
Comment (TobiasPr) [inline server/controllers/nlp.js]: please rename this variable in `tokenMatched` in this and all similiar functions

File: server/controllers/nlp.js [hunk 12]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+        state.layer = 0;
+        findCommand(state);
+    } else {
+        // this layer does not match
+        nextActionAfterNotMatched(state);
     }
 }
 
+function extractColumn(state) {
+    knex('human_resources__core_dataset').columnInfo().then(function (columnInfo) {
```
Comment (TobiasPr) [inline server/controllers/nlp.js]: state should be extended by dataset and used here

File: server/controllers/nlp.js [hunk 13]
```diff
@@ -17,19 +21,176 @@ let functions = {
...
+        var currentToken = state.tokens[state.currentToken]
+        var result = _.find(columnInfo, function (o, i) {
+            return i === currentToken;
```
Comment (TobiasPr) [inline server/controllers/nlp.js]: create ticket for including column synonyms (should be at this point)

File: server/controllers/nlp.js [hunk 14]
```diff
@@ -47,12 +208,76 @@ exports.handleInput = function (req, res) {
...
+
+        let initState = {
+            tokens: tokenizedInput,
+            layer: 0,
+            currentToken: 0,
+            callback: state => {
+                let bestMatch = {
+                    function: 'plotPieChartOfColumn',
+                    parameters: []
+                };
```
Comment (TobiasPr) [inline server/controllers/nlp.js]: bestMatch is not used, you can remove it

File: server/controllers/nlp.js [hunk 15]
```diff
@@ -17,19 +29,210 @@ let functions = {
...
+                type: 'histogram'
+            },
+            {
+                // map [{<columnName>: <columnValue>}, ...] to [<columnValue>, ...]
+                x: data.map((value) => value.attributes[column2]),
+                type: 'histogram'
+            }]
+        ))
+    },
+
```
Comment (sebmarste) [inline server/controllers/nlp.js]: please remove this empty line

File: server/controllers/nlp.js [hunk 16]
```diff
@@ -1,5 +1,17 @@
...
+var config = require('../knexfile');
+var knex = require('knex')(config);
+var _ = require('lodash');
+var columnSynonyms = require('../data/column_synonyms');
+
+let static = {
+    column: "Column",
+    operation: "Operation",
+    chartType: "ChartType"
+
```
Comment (sebmarste) [inline server/controllers/nlp.js]: please remove this empty line

File: server/controllers/nlp.js (hunk 7 above)
Comment (sebmarste) [reply server/controllers/nlp.js]: did you do some research on this?

File: server/controllers/nlp.js [hunk 17]
```diff
@@ -17,19 +29,210 @@ let functions = {
...
+            state.layer++;
+        } else {
+            state.layer = 0;
+            state.currentToken++;
+        }
     }
+    findCommand(state);
 }
 
+
```
Comment (sebmarste) [inline server/controllers/nlp.js]: empty line?

File: server/controllers/nlp.js [hunk 18]
```diff
@@ -17,19 +29,210 @@ let functions = {
...
+        state.tokens[state.currentToken] = {type: static.operation, value: currentToken};
+    }
+    nextAction(tokenMatched, state);
+}
+
+function extractColumn(state) {
+    knex('human_resources__core_dataset').columnInfo().then(function (columnInfo) {
+        let currentToken = state.tokens[state.currentToken];
+        let matched = false;
+        let tokenMatched  = _.find(columnInfo, function (o, i) {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: can we find a better naming for these two variables?

File: server/controllers/nlp.js [hunk 19]
```diff
@@ -17,19 +29,210 @@ let functions = {
...
+                state.tokens[state.currentToken] = {type: static.chartType, value: tokenMatched }
+                state.tokens.splice(state.currentToken + 1, 1);
+                matched = true;
+            }
+        }
+    }
+    nextAction(matched, state);
+}
+
+function extractConnector(state) {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: this can be removed for now, right?

File: server/controllers/nlp.js [hunk 20]
```diff
@@ -17,19 +29,210 @@ let functions = {
...
+
+}
+
+
```
Comment (sebmarste) [inline server/controllers/nlp.js]: empty line?

File: server/controllers/nlp.js [hunk 21]
```diff
@@ -47,12 +250,70 @@ exports.handleInput = function (req, res) {
...
         let dataset = req.body.dataset;
 
-        // Use NLP to find the right command
-        let command = findCommand(dataset, req.body.input);
+        let lowerCaseInput = req.body.input.toLowerCase();
+
+        var tokenizer = new natural.WordTokenizer();
+        let tokenizedInput = tokenizer.tokenize(lowerCaseInput);
+
+
```
Comment (sebmarste) [inline server/controllers/nlp.js]: empty line?

File: server/controllers/nlp.js [hunk 22]
```diff
@@ -47,12 +250,70 @@ exports.handleInput = function (req, res) {
...
-            // Send the data back to the client
-            res.send({data: data})
-        );
+
     })
-};
\ No newline at end of file
+};
+
+function callbackFunction(state, res){
```
Comment (sebmarste) [inline server/controllers/nlp.js]: rename this to `findFunction` (maybe even `findDataTransformationFunction`, but this should match the list where these functions are in)

File: server/controllers/nlp.js (hunk 19 above)
Comment (grillmann) [reply server/controllers/nlp.js]: i thought we want to extract the connectors also later, so i just left it in, but ok

File: server/controllers/nlp.js (hunk 7 above)
Comment (grillmann) [reply server/controllers/nlp.js]: Nope i think this should be done in the next iteration

Comment (sebmarste) [reply server/controllers/nlp.js]: ok, could you then create a ticket for this please?

File: server/controllers/nlp.js [hunk 23]
```diff
@@ -17,19 +28,202 @@ let functions = {
...
+function findCommand(state) {
+    if (state.currentToken < state.tokens.length && state.layer < searchFunction.length) {
+        searchFunction[state.layer](state);
+    } else {
+        state.callback(state);
     }
 }
 
+function nextAction(matched, state) {
+    if(matched) {
```
Comment (sebmarste) [inline server/controllers/nlp.js]: you could simplify the options:
```
if(matched || state.layer < searchFunction.length - 1) {
   // go to next token
   state.currentToken++;
   state.layer = 0;
} else {
   // go to next layer
   state.layer++;
}

**Pull request merged**